File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        def union(i, j):
            p_i = parents[i]
            p_j = parents[j]
            if p_i != p_j:
                for k in range(0, n):
                    if parents[k] == p_j:
                        parents[k] = p_i

        def find(k):
            pass
            
        n = len(isConnected)
        parents = [i for i in range(n)]
        for i in range(n):
            for j in range(i+1, n):
                if isConnected[i][j]:
                    union(i,j)

        #print(parents)
        return len(set(parents))

File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<bool> visited(n, false);
        int ans = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                ans++;
                dfs(isConnected, visited, i);
            } 
        }
        return ans; 
    }
private:
    void dfs(vector<vector<int>>& isConnected, vector<bool>& visited, int i) {
        if (visited[i]) return;
        visited[i] = true;
        int n = isConnected.size();
        for (int j = 0; j < n; j++) {
            if (isConnected[i][j] == 1) {
                dfs(isConnected, visited, j);
            }
        } 
    }
};